src/cmd/ksh93/sh/xec.c

```diff
@@ -3274,9 +3276,16 @@ int sh_funscope_20120720(Shell_t *shp, int argn, char *argv[], int (*fun)(void *
     }
     shp->st.cmdname = argv[0];
     // Save trap table.
-    if ((nsig = shp->st.trapmax * sizeof(char *)) > 0 || shp->st.trapcom[0]) {
-        nsig += sizeof(char *);
-        memcpy(savstak = stkalloc(shp->stk, nsig), (char *)&shp->st.trapcom[0], nsig);
+    nsig = shp->st.trapmax;
+    if (nsig > 0 || shp->st.trapcom[0]) {
+        ++nsig;
+        savsig = malloc(nsig * sizeof(char *));
+        // Contents of shp->st.trapcom may change
+        for (isig = 0; isig < nsig; ++isig) {
+            savsig[isig] = shp->st.trapcom[isig] == Empty
```

Comment:
The magic pointer `Empty` is special-cased here but not in the similar code in subshell.c which seems wrong. I would suggest removing this special-case. Instead change this line in *src/cmd/ksh93/bltins/trap.c*:

```
shp->st.trapcom[sig] = Empty;
```

to

```
shp->st.trapcom[sig] = strdup("");
```

Then all of the places that manipulate `shp->st.trapcom` can just assume every entry was dynamically allocated.